kolibri/plugins/facility/assets/src/views/users/sidePanels/AssignCoachesSidePanel.vue

```diff
@@ -1,35 +1,347 @@
+        disgardChanges$,
+        numUsersNotEligible$,
+        selectClassesLabel$,
+        assignUsersHeading$,
+        assignToAllClasses$,
+      } = bulkUserManagementStrings;
+      const { createSnackbar } = useSnackbar();
+      const { dismissAction$ } = coreStrings;
+
+      const loadUsers = async () => {
+        isLoading.value = true;
+        const users = await FacilityUserResource.fetchCollection({
+          getParams: {
+            by_ids: Array.from(props.selectedUsers).join(','),
+          },
+        });
+        facilityUsers.value = users.map(_userState);
+        isLoading.value = false;
+      };
+      loadUsers();
+
+      // Computed properties
+      const formattedClasses = computed(() => {
+        return [...props.classes]
+          .sort((a, b) => a.name.localeCompare(b.name))
+          .map(({ id, name }) => ({ id, label: name }));
+      });
+
+      const selectedUsersCount = computed(() => props.selectedUsers.size);
+
+      const hasSelectedClasses = computed(() => selectedClasses.value.length > 0);
+
+      const getSelectedClassObjects = computed(() =>
+        props.classes.filter(cls => selectedClasses.value.includes(cls.id)),
+      );
+
+      // Filter eligible users (coaches, admins, superusers)
+      const eligibleUsers = computed(() => {
+        return facilityUsers.value.filter(
+          user =>
+            props.selectedUsers.has(user.id) &&
+            (user.kind.includes(UserKinds.COACH) ||
+              user.kind === UserKinds.ADMIN ||
+              user.kind === UserKinds.SUPERUSER ||
+              user.is_superuser),
+        );
+      });
+
+      // Filter ineligible users (learners)
+      const ineligibleUsers = computed(() => {
+        return facilityUsers.value.filter(
+          user => props.selectedUsers.has(user.id) && user.kind === UserKinds.LEARNER,
+        );
+      });
+      const ineligibleUsersCount = computed(() => ineligibleUsers.value.length);
+
+      const eligibleUsersCount = computed(() => {
+        return eligibleUsers.value.length;
+      });
+
+      // Methods
+      async function handleAssign() {
+        if (!hasSelectedClasses.value) {
+          return;
+        }
+
+        isLoading.value = true;
+        showErrorWarning.value = false;
+        createdRoles.value = [];
+
+        try {
+          await assignCoachesToClasses();
+          createSnackbar(coachesAssignedNotice$());
+          showUndoModal.value = true;
+        } catch (error) {
+          showErrorWarning.value = true;
+        } finally {
+          isLoading.value = false;
+        }
+      }
+
+      async function assignCoachesToClasses() {
+        const selectedClasses = getSelectedClassObjects.value;
+        const eligibleUserIds = eligibleUsers.value.map(user => user.id);
+        if (eligibleUserIds.length === 0) {
+          return;
+        }
+
+        for (const classObj of selectedClasses) {
+          const newRoles = await RoleResource.saveCollection({
```

Comment:
(Important) We are making too many api requests with this for. We can bulk save data from multiple users and multiple classes in the same request, we don't need to make a single request for each class. Look at [this](https://github.com/ozer550/kolibri/blob/9ff8e08c51a5caff5389be762025b4d871aab542/kolibri/plugins/facility/assets/src/views/users/sidePanels/UserCreate/index.vue#L356) for example, where we use this method to save multiple classes for the same user.